diff --git a/loader/wine_info.plist.in b/loader/wine_info.plist.in
index aea99e5..4a55b5d 100644
--- a/loader/wine_info.plist.in
+++ b/loader/wine_info.plist.in
@@ -4,8 +4,6 @@
 <dict>
     <key>CFBundleDevelopmentRegion</key>
     <string>English</string>
-    <key>CFBundleExecutable</key>
-    <string>wine</string>
     <key>CFBundleIdentifier</key>
     <string>org.winehq.wine</string>
     <key>CFBundleInfoDictionaryVersion</key>
