diff --git a/include/config.h.in b/include/config.h.in
index 983c732..31381f1 100644
--- a/include/config.h.in
+++ b/include/config.h.in
@@ -1308,9 +1308,6 @@
 #undef X_DISPLAY_MISSING
 
 /* Enable large inode numbers on Mac OS X 10.5.  */
-#ifndef _DARWIN_USE_64_BIT_INODE
-# define _DARWIN_USE_64_BIT_INODE 1
-#endif
 
 /* Number of bits in a file offset, on hosts where this is settable. */
 #undef _FILE_OFFSET_BITS


