diff --git a/configure b/configure
index e8bc505..a309da5 100755
--- a/configure
+++ b/configure
@@ -11745,7 +11745,7 @@ test -n "$COREAUDIO" || enable_winecoreaudio_drv=${enable_winecoreaudio_drv:-no}
 test "x$ac_cv_member_oss_sysinfo_numaudioengines" = xyes || enable_wineoss_drv=${enable_wineoss_drv:-no}
 test "$ac_cv_header_linux_joystick_h" = "yes" || enable_winejoystick_drv=${enable_winejoystick_drv:-no}
 
-if test "x$ALSALIBS$COREAUDIOLIBS" = "x" -a \
+if test "x$ALSALIBS$COREAUDIO" = "x" -a \
         "x$ac_cv_member_oss_sysinfo_numaudioengines" != xyes -a \
         "x$with_alsa$with_coreaudio$with_oss" != xnonono
 then
diff --git a/configure.ac b/configure.ac
index b213923..306a673 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1640,7 +1640,7 @@ test "x$ac_cv_member_oss_sysinfo_numaudioengines" = xyes || enable_wineoss_drv=$
 test "$ac_cv_header_linux_joystick_h" = "yes" || enable_winejoystick_drv=${enable_winejoystick_drv:-no}
 
 dnl **** Check for any sound system ****
-if test "x$ALSALIBS$COREAUDIOLIBS" = "x" -a \
+if test "x$ALSALIBS$COREAUDIO" = "x" -a \
         "x$ac_cv_member_oss_sysinfo_numaudioengines" != xyes -a \
         "x$with_alsa$with_coreaudio$with_oss" != xnonono
 then
