diff --git a/dlls/ntdll/server.c b/dlls/ntdll/server.c
index 8a01d22..90222c3 100644
--- a/dlls/ntdll/server.c
+++ b/dlls/ntdll/server.c
@@ -742,9 +742,6 @@ static void setup_config_dir(void)
             if (!(tmp_dir = malloc( p + 1 - config_dir ))) fatal_error( "out of memory\n" );
             memcpy( tmp_dir, config_dir, p - config_dir );
             tmp_dir[p - config_dir] = 0;
-            if (!stat( tmp_dir, &st ) && st.st_uid != getuid())
-                fatal_error( "'%s' is not owned by you, refusing to create a configuration directory there\n",
-                             tmp_dir );
             free( tmp_dir );
         }
 
@@ -842,7 +839,6 @@ static int server_connect(void)
 
     /* make sure we are at the right place */
     if (stat( ".", &st ) == -1) fatal_perror( "stat %s", serverdir );
-    if (st.st_uid != getuid()) fatal_error( "'%s' is not owned by you\n", serverdir );
     if (st.st_mode & 077) fatal_error( "'%s' must not be accessible by other users\n", serverdir );
 
     for (retry = 0; retry < 6; retry++)
@@ -864,8 +860,6 @@ static int server_connect(void)
         /* make sure the socket is sane (ISFIFO needed for Solaris) */
         if (!S_ISSOCK(st.st_mode) && !S_ISFIFO(st.st_mode))
             fatal_error( "'%s/%s' is not a socket\n", serverdir, SOCKETNAME );
-        if (st.st_uid != getuid())
-            fatal_error( "'%s/%s' is not owned by you\n", serverdir, SOCKETNAME );
 
         /* try to connect to it */
         addr.sun_family = AF_UNIX;
diff --git a/libs/wine/config.c b/libs/wine/config.c
index 035c268..de1b9f1 100644
--- a/libs/wine/config.c
+++ b/libs/wine/config.c
@@ -293,10 +293,6 @@ static void init_paths(void)
         }
     }
     if (!S_ISDIR(st.st_mode)) fatal_error( "%s is not a directory\n", config_dir );
-#ifdef HAVE_GETUID
-    if (st.st_uid != getuid()) fatal_error( "%s is not owned by you\n", config_dir );
-#endif
-
     init_server_dir( st.st_dev, st.st_ino );
 }
 
diff --git a/server/request.c b/server/request.c
index 05c7464..c2139b6 100644
--- a/server/request.c
+++ b/server/request.c
@@ -545,7 +545,6 @@ static void create_dir( const char *name, struct stat *st )
         if (lstat( name, st ) == -1) fatal_perror( "lstat %s", name );
     }
     if (!S_ISDIR(st->st_mode)) fatal_error( "%s is not a directory\n", name );
-    if (st->st_uid != getuid()) fatal_error( "%s is not owned by you\n", name );
     if (st->st_mode & 077) fatal_error( "%s must not be accessible by other users\n", name );
 }
 
