diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index 6875144..563c366 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -51,7 +51,7 @@ int locale_id__Id_cnt = 0;
 
 /* ?_Clocptr@_Locimp@locale@std@@0PAV123@A */
 /* ?_Clocptr@_Locimp@locale@std@@0PEAV123@EA */
-locale__Locimp *locale__Locimp__Clocptr;
+locale__Locimp *locale__Locimp__Clocptr = NULL;
 
 static const vtable_ptr MSVCP_locale_facet_vtable[];
 
