From 8f21b352c25b0078c4cb8a7c554dd8503a9d136f Mon Sep 17 00:00:00 2001
From: Nikolay Sivov <nsivov@codeweavers.com>
Date: Wed, 25 Jan 2017 22:51:32 +0300
Subject: [PATCH] shell32: Fix return value of GetDetailsOf() for unixfs
 folder.

Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>
Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 dlls/shell32/shfldr_unixfs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/shell32/shfldr_unixfs.c b/dlls/shell32/shfldr_unixfs.c
index 9df761f..3feef74 100644
--- a/dlls/shell32/shfldr_unixfs.c
+++ b/dlls/shell32/shfldr_unixfs.c
@@ -1475,10 +1475,10 @@ static HRESULT WINAPI ShellFolder2_GetDetailsOf(IShellFolder2* iface,
     LPCITEMIDLIST pidl, UINT iColumn, SHELLDETAILS *psd)
 {
     UnixFolder *This = impl_from_IShellFolder2(iface);
-    HRESULT hr = E_FAIL;
     struct passwd *pPasswd;
     struct group *pGroup;
     struct stat statItem;
+    HRESULT hr = S_OK;
 
     static const shvheader unixfs_header[SHELLVIEWCOLUMNS] = {
         {IDS_SHV_COLUMN1,  SHCOLSTATE_TYPE_STR  | SHCOLSTATE_ONBYDEFAULT, LVCFMT_RIGHT, 15},
