--- a/dlls/user32/winstation.c
+++ b/dlls/user32/winstation.c
@@ -462,9 +462,18 @@
  */
 HDESK WINAPI OpenInputDesktop( DWORD flags, BOOL inherit, ACCESS_MASK access )
 {
-    FIXME( "(%x,%i,%x): stub\n", flags, inherit, access );
-    SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
-    return 0;
+    HANDLE ret = 0;
+
+    SERVER_START_REQ(open_input_desktop)
+    {
+       req->flags = flags;
+       req->access = access;
+       req->attributes = OBJ_CASE_INSENSITIVE | (inherit ? OBJ_INHERIT : 0);
+       if(!wine_server_call( req )) ret = wine_server_ptr_handle( reply->handle );
+    }
+    SERVER_END_REQ;
+
+    return ret;
 }
