--- a/server/user.h
+++ b/server/user.h
@@ -49,6 +49,7 @@
     struct list        desktops;           /* list of desktops of this winstation */
     struct clipboard  *clipboard;          /* clipboard information */
     struct atom_table *atom_table;         /* global atom table */
+    struct desktop    *active_desktop;     /* Active Desktop of this winstation */
 };
 
 struct global_cursor
@@ -182,6 +183,7 @@
                                          obj_handle_t handle );
 extern void close_process_desktop( struct process *process );
 extern void close_thread_desktop( struct thread *thread );
+extern int is_active_desktop( struct desktop *desktop );
 
 /* mirror a rectangle respective to the window client area */
 static inline void mirror_rect( const rectangle_t *client_rect, rectangle_t *rect )
