diff --git a/tools/winegcc/winegcc.c b/tools/winegcc/winegcc.c
index fe0f033..03cc131 100644
--- a/tools/winegcc/winegcc.c
+++ b/tools/winegcc/winegcc.c
@@ -1240,11 +1240,11 @@ static int is_target_arg(const char* arg)
 
 /*
  *      Directory Options
- *          -Bprefix  -Idir  -I-  -Ldir  -specs=file
+ *          -Bprefix  -Idir  -I-  -Ldir  -specs=file, -isysrootfile, -iframeworkfile
  */
 static int is_directory_arg(const char* arg)
 {
-    return arg[1] == 'B' || arg[1] == 'L' || arg[1] == 'I' || strncmp("-specs=", arg, 7) == 0;
+    return arg[1] == 'B' || arg[1] == 'L' || arg[1] == 'I' || strncmp("-specs=", arg, 7) == 0 || strncmp("-isysroot", arg, 8) == 0 || strncmp("-iframework", arg, 11) == 0;
 }
 
 /*
