diff --git a/dlls/winemac.drv/cocoa_app.m b/dlls/winemac.drv/cocoa_app.m
index d454513..eb29ade 100644
--- a/dlls/winemac.drv/cocoa_app.m
+++ b/dlls/winemac.drv/cocoa_app.m
@@ -215,7 +215,7 @@ - (void) transformProcessToForeground
             mainMenu = [[[NSMenu alloc] init] autorelease];
 
             // Application menu
-            submenu = [[[NSMenu alloc] initWithTitle:@"Wine"] autorelease];
+            submenu = [[[NSMenu alloc] initWithTitle:@"PlayOnMac"] autorelease];
             bundleName = [[NSBundle mainBundle] objectForInfoDictionaryKey:(NSString*)kCFBundleNameKey];
 
             if ([bundleName length])
@@ -236,7 +236,7 @@ - (void) transformProcessToForeground
             else
                 title = @"Quit";
             item = [submenu addItemWithTitle:title action:@selector(terminate:) keyEquivalent:@"q"];
-            [item setKeyEquivalentModifierMask:NSCommandKeyMask | NSAlternateKeyMask];
+            [item setKeyEquivalentModifierMask:NSCommandKeyMask];
             item = [[[NSMenuItem alloc] init] autorelease];
             [item setTitle:@"Wine"];
             [item setSubmenu:submenu];
