From c98a3c976f911915755118b894fce783981ff851 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Tue, 17 Mar 2015 00:27:45 +0100
Subject: server: Replace strndup with equivalent sequence of mem_alloc and
 memcpy.

---
 server/file.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/server/file.c b/server/file.c
index 7736ea5..9d6d3cb 100644
--- a/server/file.c
+++ b/server/file.c
@@ -444,8 +444,9 @@ static struct security_descriptor *file_get_parent_sd( struct fd *root, const ch
     struct stat st;
     int unix_fd;
 
-    parent_name = strndup( child_name, child_len );
-    if (!parent_name) return NULL;
+    if (!(parent_name = mem_alloc( child_len + 1 ))) return NULL;
+    memcpy( parent_name, child_name, child_len );
+    parent_name[child_len] = 0;
 
     /* skip trailing slashes */
     p = parent_name + strlen( parent_name ) - 1;
-- 
2.3.2
